python/ray/data/datasource/huggingface_datasource.py

```diff
@@ -0,0 +1,76 @@
+from typing import TYPE_CHECKING, Iterable, List, Optional, Union
+
+from ray.data._internal.dataset_logger import DatasetLogger
+from ray.data._internal.util import _check_pyarrow_version
+from ray.data.block import Block, BlockMetadata
+from ray.data.datasource import Datasource, Reader, ReadTask
+from ray.util.annotations import DeveloperAPI
+
+logger = DatasetLogger(__name__)
+
+if TYPE_CHECKING:
+    import datasets
+
+
+@DeveloperAPI
+class HuggingFaceDatasource(Datasource):
+    """Hugging Face Dataset datasource, for reading from a
+    `Hugging Face Datasets Dataset <https://huggingface.co/docs/datasets/package_reference/main_classes#datasets.Dataset/>`_.
+    This Datasource implements a streamed read, most beneficial for a
+    `Hugging Face Datasets IterableDataset <https://huggingface.co/docs/datasets/package_reference/main_classes#datasets.IterableDataset/>`_
+    or datasets which are too large to fit in-memory.
+    For an in-memory Hugging Face Dataset (`datasets.Dataset`), use :meth:`~ray.data.from_huggingface`
+    for faster performance.
+    """  # noqa: E501
+
+    def create_reader(
+        self,
+        dataset: Union["datasets.Dataset", "datasets.IterableDataset"],
+    ) -> "_HuggingFaceDatasourceReader":
+        return _HuggingFaceDatasourceReader(dataset)
+
+
+class _HuggingFaceDatasourceReader(Reader):
+    def __init__(
+        self,
+        dataset: Union["datasets.Dataset", "datasets.IterableDataset"],
+        batch_size: int = 4096 * 64,
+    ):
+        self._dataset = dataset
+        self._batch_size = batch_size
+
+    def estimate_inmemory_data_size(self) -> Optional[int]:
+        return self._dataset.dataset_size
+
+    def get_read_tasks(
+        self,
+        parallelism: int,
+    ) -> List[ReadTask]:
+        # Note that `parallelism` arg is currently not used for HuggingFaceDatasource.
```

Comment:
I thought we wanted to use `split_dataset_by_node` so that we can have distributed reads for the streaming case?

Reply:
Yeah that was the initial intent of the PR. We did further investigation into `split_dataset_by_node`, and turned out that it doesn't actually shard the dataset. It actually reads in the same dataset for each node, and selects a subset to emulate sharding -- so it doesn't provide any efficiency gains. That's why this PR only uses a single read task because we cannot distribute the dataset read.

Reply:
Looking at the code, it seems there is logic to shard the base IterableDataset if it contains multiple files:

https://sourcegraph.com/github.com/huggingface/datasets/-/blob/src/datasets/iterable_dataset.py?L1231-1243
https://github.com/huggingface/datasets/issues/5984

Only if the base dataset contains a single file, does it not get sharded

Reply:
Example:

```
>>> hf_ds = datasets.load_dataset("openclimatefix/gfs-surface-pressure-2.0deg", split='train', streaming=True)
Downloading metadata: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 14.5k/14.5k [00:00<00:00, 4.98MB/s]
Using custom data configuration openclimatefix--gfs-surface-pressure-2.0deg-e3bd919c6fc2ba90
>>> print(hf_ds.n_shards)
39
```

Reply:
If `n_shards` is >= `parallelism` then we should be able to do a proper distributed read